### ISSUE TITLE
Picky Image Loading

### ISSUE THREAD

**Issue by alanspires** [luberda-molinet/FFImageLoading#387]
I can consistently get a image to NOT load on startup, even though the image is disk cached and moving the list view to redraw the cell loads it. You can see in the video that the image, using the same URL, is correctly displayed just two cells underneath it as well.

https://www.dropbox.com/s/wx9sghxxj64jsio/PickyImage.mov?dl=0 

Really weird, any thoughts?

**Comment by rusty21** [comment]
The first place I would start is making sure you are on the latest version. Next look at the application output to see FFImageLoading printed some error or stack trace. If there's nothing there I would probably implement a logger for FFImageLoading. Mine looks like this

	public class MyLogger : IMiniLogger
	{
		public void Debug (string message)
		{
			System.Diagnostics.Debug.WriteLine (message);
		}

		public void Error (string errorMessage)
		{
			System.Diagnostics.Debug.WriteLine (errorMessage);
		}

		public void Error (string errorMessage, Exception ex)
		{
			System.Diagnostics.Debug.WriteLine (errorMessage + " " + ex.ToString());
		}
	}

You use it like this. 

    ImageService.Instance.Config.Logger = new MyLogger ();

I actually ran into something similar hopefully this will work for you too. In my tableviews in order to not see images where they don't belong I was doing something like this in my GetCell method for my source

    imageTask?.Cancel();
    imgImage.Image = UIImage.FromBundle(...);
    imageTask = ImageService.LoadUrl(urlToBeLoaded).Into();

After doing this I saw that sometimes images loaded from the same url wouldn't load into their images views. I ended modifying the code a bit like this:

    if(currentUrlInImageView != urlToBeLoaded)
    {
        imageTask?.Cancel();
    }
    imgImage.Image = UIImage.FromBundle(...);
    imageTask = ImageService.LoadUrl(urlToBeLoaded).Into();

Basically only cancel the previous task if I'm trying to load a different image than the one that is already loading into the imageview. I'm not sure if this effects you or can help because you already are seeing failures on load without scrolling but who knows.

**Comment by alanspires** [comment]
Thanks, I'll check that out. I don't have any errors right now and I've been on the latest pre-release.

**Comment by alanspires** [comment]
@rusty21  looks like your working in native, I'm in Forms and not using custom renderers for these ListViews.

**Comment by daniel-luberda** [comment]
@rusty21 @alanspires 
In forms, previous image loading is cancelled by default. 

@alanspires I can't reproduce it, any chance for reproduction? Thanks

**Comment by daniel-luberda** [comment]
@alanspires Any news?

**Comment by alanspires** [comment]
Here is a link that shows the run with logging. It appears that the image gets queued but the download never happens until scroll.

https://www.dropbox.com/s/0qn81kil3l42prc/PickyImageLoadingWithDebug.mov?dl=0

(Sorry gonna try with logging info before i attempt to create a repo)

Does this help?

**Comment by alanspires** [comment]
Same on new pre

**Comment by alanspires** [comment]
Further investigation shows that

https://github.com/luberda-molinet/FFImageLoading/blob/0b6c156a3e6cd967185924094f7173e8af0d3575/source/FFImageLoading.Common/Work/WorkScheduler.cs#L274

never returns on certain images which causes the image not to load. I'm going to keep debugging.

**Comment by alanspires** [comment]
Here is a repo

https://github.com/alanspires/FFImageBugs

I also noticed that this only occurs when "HasUnevenRows=true". Launch the project and you will see that there is a image that doesn't load. Get rid of HasUnevenRows and set a RowHeight and the problem stops.

**Comment by CliffCawley** [comment]
Just adding that I've experienced the same issue as @alanspires. Thanks for the tip on how to overcome, I just can't use it for most of my lists since the content is dynamically sized vertically. I've been using 2.2.5

**Comment by alanspires** [comment]
@CliffCawley all of mine are dynamic as well, I'm confident that when @daniel-luberda gets back in the mix he can sort this one out.

**Comment by alanspires** [comment]
Oops mis read that, from debugging the issue seems to be related to the use case of the same image being requested multiple times and the deduplication code. Confident Daniel will sort this out quickly when he gets back.

**Comment by alanspires** [comment]
@daniel-luberda Any thoughts on this and the repo?

**Comment by daniel-luberda** [comment]
Hi @alanspires, @CliffCawley,

I tested reproduction (many thanks for that @alanspires). Conclusions are:

- It's probably not an FFImageLoading bug, rather Xamarin.Forms layout system issue
- Image loading is done in every case (correct `UIImage` is assigned to `UIImageView` control every time), but with `HasUnevenRows=true` it's not reflected on a cell layout (but the `UIImage` is correctly assigned as stated before). 
- The same list with `HasUnevenRows=false` works just fine

Any help with that?

**Comment by alanspires** [comment]
I'm not sure this is the case. I have a app store version running on FFImage 2.1.8 and this issue never occurs.

**Comment by alanspires** [comment]
I updated the sample repo for 2.1.8 and it works flawless now. This would lead me to think it is a issue with the library.

https://github.com/alanspires/FFImageBugs

**Comment by alanspires** [comment]
@daniel-luberda any news on this? I mean this almost makes the library un-usable in production. Having a ListView with "HasUnevenRows" is such a basic scenario. Not to mention this problem does not exists in older versions of the lib.

Also I don't think the "invalid" tag is right.

**Comment by daniel-luberda** [comment]
@alanspires That was so tricky... See this: https://github.com/luberda-molinet/FFImageLoading/commit/16a3cfc7a012945f01804e59441adaac32aeef2c

I didn't test your example yet, but that should be it. It also should fix a lot of other Xamarin.Forms layout issues.

**Comment by daniel-luberda** [comment]
Could you test the newest alpha?

**Comment by alanspires** [comment]
Good find, but doesn't fix this issue.

**Comment by daniel-luberda** [comment]
Hmmm. Do you know if it also affects Android? I'm really puzzled with that. Why would something related to layout (`HasUnevenRows`) cause that? That could be tricky.

**Comment by daniel-luberda** [comment]
Did you use Xamarin.FFImageLoading.2.2.6-pre-258 nuget for testing?

**Comment by alanspires** [comment]
Yes I used 258. Why would version 2.1.8 not be affected by this?

**Comment by daniel-luberda** [comment]
@alanspires I'll try to debug it once again. 

> Why would version 2.1.8 not be affected by this?

2.2 is quite faster in some scenarios, maybe it updates image before some XF layout calculations, etc.

**Comment by alanspires** [comment]
@daniel-luberda appreciate the effort. The first impression when my app opens is placeholder images that never load until you scroll them out of view and back in. This just seems like such a basic problem that shouldn't be happening. Thanks!

**Comment by daniel-luberda** [comment]
After debugging, I am quite sure that it's not an **FFImageLoading** issue but **Xamarin.Forms** issue. Here's why: 

- There aren't any issues on Android (it's an iOS specific issue)
- It works good when `ViewCell`'s code is replaced with: 

```C#
public TestCell()
{
	_image = new CachedImage
	{
		Aspect = Aspect.AspectFill,
		DownsampleToViewSize = true,
		CacheDuration = TimeSpan.FromDays(30),
		RetryCount = 10,
		RetryDelay = 10,
		LoadingPlaceholder = "loading.png",
		ErrorPlaceholder = "error.png"
	};

	// OLD CODE
	// View = new StackLayout() { Children = { _image } };

	var label = new Label();
	label.SetBinding(Label.TextProperty, "Description");

	var root = new Grid();
	root.Children.Add(_image);
	root.Children.Add(label);

	View = root;
}
```

The issue is clearly with having `StackLayout` inside `ViewCell` when `HasUnevenRows` is set to `true`.  IMO it should be reported as a Xamarin.Forms issue.

**Comment by alanspires** [comment]
My production code that has this issue uses a grid.

**Comment by daniel-luberda** [comment]
I don't have any idea how could we fix that as:

- The `UIImageView` inside cell is updated with correct `UIImage`.
- We use the same method to notify Xamarin.Forms about view size changes, see: 
  - [Image](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.iOS/Renderers/ImageRenderer.cs#L128)
  - [CachedImage](https://github.com/luberda-molinet/FFImageLoading/blob/master/source/FFImageLoading.Forms.Touch/CachedImageRenderer.cs#L316)
- It just works when `HasUnevenRows` is set to `false`

Any ideas?

**Comment by daniel-luberda** [comment]
The problems you had when using `Grid` in you project may be related to another issue which was fixed recently: https://github.com/luberda-molinet/FFImageLoading/commit/16a3cfc7a012945f01804e59441adaac32aeef2c

**Comment by alanspires** [comment]
If that fix is in the latest pre-release it didn't solve it for me. Hmmm. this is tricky.

**Comment by daniel-luberda** [comment]
It gets even weirder:

![screen shot 2016-12-28 at 19 57 25](https://cloud.githubusercontent.com/assets/13569983/21529164/ebb83480-cd37-11e6-8ade-7b5f9d250cd9.png)

**Comment by alanspires** [comment]
Wow I never caught that one, now I'm even more confused lol.

**Comment by daniel-luberda** [comment]
I am also confused. It's definitely related to Xamarin.Forms layout system, but... I don't know where to start. :)

**Comment by daniel-luberda** [comment]
@alanspires  @rusty21  I found some issues in WorkScheduler which caused images not loading under some scenarios :) It's fixed in the newest prerelease (FFImageLoading.2.2.7-pre-294). Please reopen if you still experience it. Thanks.

**Comment by alanspires** [comment]
@daniel-luberda awesome! I'll check it out
